README.md
# SnailFM
[![Snap](https://snapcraft.io/snailfm/badge.svg)](https://snapcraft.io/snailfm)
SnailFM is a radio station by the SnailDOS community.
You can easily tune into SnailFM with this software.
SnailFM is a online radiostation, with no copyright music. It has no lyrics, so it's great for background music, or just studying or even coding!
SnailFM is fast free and secure!
For linux installation, follow the guide below.

# INSTALLATION
Follow the WIKI!
https://github.com/snaildos/SnailFM-Application/wiki/Installation

# Building by source
You can easily build SnailFM. I recommend linux.
`yarn` 
`yarn start`
then building can be:
`yarn build` (linux only)

# Contributing
Feel free to contribute! Fork, commit, PR!

[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fsnaildos%2FSnailFM-Application.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fsnaildos%2FSnailFM-application?ref=badge_large)


CHANGELOG.md
# Changelog

<a name="2.1.8"></a>
## 2.1.8 (2023-07-06)

### Miscellaneous

-  BUMP: Minor packages [[27d4daf](https://github.com/snaildos/SnailFM-Application/commit/27d4daf0c97ea31ae0240a9d26b6338f84db2d43)]
-  FEAT: Rewrite compatibility support [[b16dc25](https://github.com/snaildos/SnailFM-Application/commit/b16dc25cea337634a5d2a97f9970ff816f9ce9dd)]
-  Merge pull request [#109](https://github.com/snaildos/SnailFM-Application/issues/109) from snaildos/renovate/electron-25.x [[4a58f2d](https://github.com/snaildos/SnailFM-Application/commit/4a58f2d158c5ebf93aa3fb4b067cfb675c93b9f0)]
-  Merge pull request [#106](https://github.com/snaildos/SnailFM-Application/issues/106) from snaildos/renovate/axios-1.x [[ace7ef5](https://github.com/snaildos/SnailFM-Application/commit/ace7ef50915ee670acddda7f0d9115ceb739294e)]
-  Merge pull request [#108](https://github.com/snaildos/SnailFM-Application/issues/108) from snaildos/renovate/jquery-3.x [[c59679d](https://github.com/snaildos/SnailFM-Application/commit/c59679de0fa7441759cc974205326e1886096501)]
-  Merge pull request [#110](https://github.com/snaildos/SnailFM-Application/issues/110) from snaildos/renovate/electron-builder-24.x [[4b39fb0](https://github.com/snaildos/SnailFM-Application/commit/4b39fb0416206bb5fe579f12d5b3e57e93f18348)]
-  Merge pull request [#111](https://github.com/snaildos/SnailFM-Application/issues/111) from snaildos/renovate/electron-updater-6.x [[9d5cdec](https://github.com/snaildos/SnailFM-Application/commit/9d5cdec872b56b4acf2df1800d53c0a6dc7168f8)]
-  Update dependency electron to v25 [[b466cb0](https://github.com/snaildos/SnailFM-Application/commit/b466cb0d220a0007bec118c27196722fd62cee5d)]
-  Update dependency electron-updater to v6 [[862a107](https://github.com/snaildos/SnailFM-Application/commit/862a1070e5ee0ad6128a241d5db706152e9b6eb8)]
-  Update dependency electron-builder to v24 [[9f44875](https://github.com/snaildos/SnailFM-Application/commit/9f44875b8b74cb27a72754de6dd8cbbe35a15f56)]
-  Update dependency jquery to v3.7.0 [[a7c277b](https://github.com/snaildos/SnailFM-Application/commit/a7c277b05a6205ffbd659dfd9b910950bdbd0720)]
-  Update dependency axios to v1.4.0 [[697a032](https://github.com/snaildos/SnailFM-Application/commit/697a032aa3e64c660e58021bfcaace438d7cc53a)]
-  FEAT: Cleanup [[8eb0b15](https://github.com/snaildos/SnailFM-Application/commit/8eb0b1509161daf89af5fe0adabaf63899046dca)]
-  Update CHANGELOG.md [[2d8ce0e](https://github.com/snaildos/SnailFM-Application/commit/2d8ce0e4f432ac0895490d1f5349f559daf2a6b4)]
-  Merge pull request [#105](https://github.com/snaildos/SnailFM-Application/issues/105) from snaildos/fix_transparent [[aabe231](https://github.com/snaildos/SnailFM-Application/commit/aabe231e3c414853b19a009aaba87ac786a39dfd)]
-  Update nonet.html [[f885658](https://github.com/snaildos/SnailFM-Application/commit/f8856587b1bb38c8fa3196cde4736ff3108fbbc7)]
-  Update logo imports [[312037c](https://github.com/snaildos/SnailFM-Application/commit/312037ce05179e483c9721e4b622e8ce8d8e164d)]
-  Rewrite index.html to make everything cleaner [[9fa9ceb](https://github.com/snaildos/SnailFM-Application/commit/9fa9ceb4ed3ac0d7309468dd960867f0856ff4f0)]
-  Update update.html [[1883be1](https://github.com/snaildos/SnailFM-Application/commit/1883be196b0907be274a7f4635746136d371c6ae)]
-  Update Transparency module [[64759e3](https://github.com/snaildos/SnailFM-Application/commit/64759e34ee4d3c2547fb032628cc4defc5f00357)]
-  Merge branch &#x27;main&#x27; of https://github.com/snaildos/SnailFM-Application [[1f50c8f](https://github.com/snaildos/SnailFM-Application/commit/1f50c8f4944490ad7b49541f67533daaea9588ec)]
-  Update package.json [[a4fb791](https://github.com/snaildos/SnailFM-Application/commit/a4fb791b4b4c2abc571cc068fdac037f983e606a)]
-  Merge pull request [#103](https://github.com/snaildos/SnailFM-Application/issues/103) from snaildos/renovate/electron-23.x [[4f49f47](https://github.com/snaildos/SnailFM-Application/commit/4f49f479dc8d6858a82db838df7c3401b9fd0266)]
-  Merge pull request [#104](https://github.com/snaildos/SnailFM-Application/issues/104) from snaildos/renovate/utf-8-validate-6.x [[e3b7b4e](https://github.com/snaildos/SnailFM-Application/commit/e3b7b4e720b27dfa0c847a7dd471e299b3a18f0c)]
-  Merge pull request [#101](https://github.com/snaildos/SnailFM-Application/issues/101) from snaildos/renovate/jquery-3.x [[ecd3716](https://github.com/snaildos/SnailFM-Application/commit/ecd3716fc850e7720c0900bfe0f87304882438cb)]
-  Merge pull request [#102](https://github.com/snaildos/SnailFM-Application/issues/102) from snaildos/renovate/axios-1.x [[c311ef6](https://github.com/snaildos/SnailFM-Application/commit/c311ef6130bedfdf0379146b47a6e87ed17a58f4)]
-  Update dependency utf-8-validate to v6 [[719aaaf](https://github.com/snaildos/SnailFM-Application/commit/719aaaf842127610b4627d1ee150d1725318ed2d)]
-  Update dependency electron to v23 [[b02027c](https://github.com/snaildos/SnailFM-Application/commit/b02027ce7ef27b57647f27363d78f854f28b090a)]
-  Update dependency axios to v1.3.4 [[029fef5](https://github.com/snaildos/SnailFM-Application/commit/029fef591b4d9066d952175e634047c41093e0b8)]
-  Update dependency jquery to v3.6.4 [[3ed9a77](https://github.com/snaildos/SnailFM-Application/commit/3ed9a774de4e300300c748ad1d37b05ec5b15650)]
-  Merge pull request [#96](https://github.com/snaildos/SnailFM-Application/issues/96) from snaildos/renovate/electron-builder-23.x [[02c9c79](https://github.com/snaildos/SnailFM-Application/commit/02c9c79c0cf64efd7261e3a9b152c62c29bf3ea6)]
-  Merge pull request [#97](https://github.com/snaildos/SnailFM-Application/issues/97) from snaildos/renovate/electron-updater-5.x [[1d64e13](https://github.com/snaildos/SnailFM-Application/commit/1d64e133eced358be25b77b31b75ee9b11ad9bf3)]

main.js
const { autoUpdater } = require("electron-updater");
const { SSL_OP_EPHEMERAL_RSA } = require("constants");
const { app, BrowserWindow, ipcMain, protocol, ipcRenderer } = require("electron");
const { watchFile } = require("fs");
const { trackEvent } = require("./lib/analytics.js");
const {TitlebarRespect} = require('electron-titlebar-respect')
const glasstron = require('glasstron-clarity');
const electron = require("electron");
const isOnline = require('is-online');
const Store = require("electron-store");
const store = new Store();
const axios = require('axios');
token = store.get('token')
store.set('createonlogin', false)
// Notify
const { Notification } = require("electron");
electron.app.commandLine.appendSwitch("enable-transparent-visuals");
TitlebarRespect({})

// wait function
function wait(ms) {
  var d = new Date();
  var d2 = null;
  do {
    d2 = new Date();
  } while (d2 - d < ms);
}

// Start the libaries
require("./lib/rpc.js");
console.log("RPC lib init.");


// Loading screen
/// create a global var, wich will keep a reference to out loadingScreen window
let loadingScreen;
const createLoadingScreen = () => {
  loadingScreen = new BrowserWindow(
    Object.assign({
      width: 700,
      height: 120,
      alwaysOnTop: true,
      frame: false,
      fullscreen: false,
      show: true,
      transparent: true,
      webPreferences: {
        sandbox: true,
        nodeIntergration: false,
        contextIsolation: true
      }
    })
  );
  loadingScreen.setResizable(false);
  loadingScreen.loadFile("./src/pages/splash.html");
  loadingScreen.on("closed", () => (loadingScreen = null));
  loadingScreen.webContents.on("did-finish-load", () => {
    loadingScreen.show();
  });
};
console.log("Loading screen ready.");

// Start the main program
let mainWindow;

function createWindow() {
  mainWindow = new glasstron.BrowserWindow({
    width: 800,
    height: 600,
    show: false,
    frame: global.frame,
    titleBarStyle: global.titleBarStyle,
    fullscreen: false,
    blur: true,
    blurType: global.blurType,
    modal: true,
    icon: "snailfm.ico",
    trafficLightPosition: { x: 24, y: 25 },
    webPreferences: {
      nodeIntegration: true,
      nativeWindowOpen: true,
      contextIsolation: false,
    },
  });
  mainWindow.setMenuBarVisibility(false);
  mainWindow.setResizable(false);
  mainWindow.loadFile("index.html");
  mainWindow.on("maximize", () => mainWindow.unmaximize());
  ipcMain.on("main_show", () => mainWindow.show());
  ipcMain.on('main_hide', () => {mainWindow.hide();})
  ipcMain.on("load_acc", () => mainWindow.loadFile('myaccount.html'))
  mainWindow.webContents.on("did-finish-load", () => {
    if (loadingScreen) {
      loadingScreen.close();
      trackEvent("stnailfm.started", "SnailFM has just loaded up!");
    }
    var isDev = require("isdev");

    if (isDev) {
      console.log("In Development!");
    } else {
      console.log("Not in Development!");
    }


    (async () => {
      if ((await isOnline() === false)) {
        console.log("No internet");
        const notification3 = {
          title: "SnailFM",
          body: "No valid network connection! Please reconnect!",
        };
      mainWindow.loadFile("./src/pages/nonet.html");
      new Notification(notification3).show();
      }
    })();

    mainWindow.show();
    console.log("Ok! Window init, let's check for updates...");
    autoUpdater.checkForUpdatesAndNotify();
    console.log("Update checked. Let's see what happens!");
  });

  mainWindow.on("closed", function () {
    ipcMain.emit('login_exit');
  });
}

console.log("Main screen ready.");
console.log("Setup login function...");
var loginclose = 0
function createloginWindow() {
  const loginWindow = (new BrowserWindow({ 
    width: 1200,
    height: 700,
    minWidth: 800,
    minHeight: 400,
    center: true,
    show: false,
    webPreferences: {
      nodeIntegration: false,
      contextIsolation: true,
    }
  }));
  loginWindow.hide()
  loginWindow.on('close', function (event) {
    event.preventDefault();
    const notification4 = {
      title: "SnailFM",
      body: "This window won't close unless you login! If you want to exit SnailFM, press the close button 3 times!",
    };
    loginclose = loginclose + 1;
    if (loginclose === 3) {
      store.set('logged_in', false)
      console.log("App quitting...")
      ipcMain.emit('appquit')
    } else {
      if (loginclose === 1) {
        new Notification(notification4).show();
      }
    }

  });
  loginWindow.setMenuBarVisibility(false) 
  ipcMain.on('login_sar', () => {loginWindow.show(); loginWindow.loadURL("https://login.snaildos.com?redir=snailfm://login"); loginWindow.maximize(); ipcMain.emit('main_hide')})
  ipcMain.on('login_show', () => {loginWindow.show(); loginWindow.maximize(); loginWindow.restore()})
  ipcMain.on('login_hide', () => {loginWindow.hide();})
  ipcMain.on('login_exit', () => {loginWindow.destroy();})
  loginWindow.loadURL("https://login.snaildos.com?redir=snailfm://login");
}

app.whenReady().then(() => {
  ipcMain.on('token_verify', function() {
    if ((store.get('logged_in') === false)) {
      console.log("All good! Login is disabled...");
      createWindow()
    } else {
  if (token != null) {
    console.log("Verifying Identity..")
    async function verify() {
      await axios.get('https://api.snaildos.com/users/@me', {
      headers: {
      Authorization: token
      }}
      )
        .then(function (response) {
          // handle success
          if (( response.data.status ) === "false") {
            config.set('token', null)
            console.log("Token failure!")
            ipcMain.emit("login_sar")
            loadingScreen.close();
            store.set('createonlogin', true)
          } else {
            console.log("Sucess!")
            createWindow();
          }
        })
        .catch(function (error) {
          // handle error
          console.log(error);
          ipcMain.emit("login_sar")
          console.log("Token failure!")
          loadingScreen.close();
          store.set('createonlogin', true)
        })
        .then(function () {
        });
      }
      verify()
  } else {
    if ((store.get('logged_in') === false)) {
      console.log("All good! Login is disabled...");
      createWindow()
    } else {
      if ((store.get('setup') != undefined)) {
      console.log("User has login enabled but is not logged in, showing login screen...");
      ipcMain.emit("login_sar")
      console.log("Token failure!")
      loadingScreen.close();
      } else {
        createWindow()
      }
    }
  }
}
})

  app.setAsDefaultProtocolClient("snailfm");
  protocol.registerHttpProtocol('snailfm', (req, cb) => {
    const url = req.url.substr(22)
      console.log("Logging in: "+url);
      console.log("Token data: "+url)
      store.set('token', url);
      ipcMain.emit('load_acc');
      ipcMain.emit('login_hide')
      ipcMain.emit('main_show')
      store.set('logged_in', true)
      if ((store.get('createonlogin')) === true) {
        createWindow()
        store.set('createonlogin', false);
      }
  })

  app.on("open-url", (event, url) => {
    const url2 = req.url.substr(22)
    console.log("Logging in: "+url2);
    console.log("Token data: "+url2)
    store.set('token', url2);
    ipcMain.emit('load_acc');
    ipcMain.emit('login_hide')
    ipcMain.emit('main_show')
    store.set('logged_in', true);
    if ((store.get('createonlogin')) === true) {
      createWindow()
      store.set('createonlogin', false);
    }
  });
});


app.on("ready", () => {
  createLoadingScreen();
  console.log("Send check for updates signal...");
  console.log("Alright, lets go!");
  setTimeout(() => {
    createloginWindow()
    ipcMain.emit('token_verify');
  }, 3000);
});

app.on("window-all-closed", function () {
  if (process.platform !== "darwin") {
    app.quit();
  }
});

ipcMain.on("app_version", (event) => {
  event.sender.send("app_version", { version: app.getVersion() });
});

autoUpdater.on("update-available", () => {
  mainWindow.webContents.send("update_available");
});

autoUpdater.on("update-downloaded", () => {
  mainWindow.webContents.send("update_downloaded");
});

ipcMain.on("relaunch", () => {
  ipcMain.emit("close")
});

ipcMain.on("restart_app", () => {
  autoUpdater.quitAndInstall();
});

ipcMain.on('appquit', () => {app.exit()})
ipcMain.on('minimize', () => {mainWindow.minimize()})
ipcMain.on('maximize', () => {mainWindow.maximize()})
ipcMain.on('restore', () => {mainWindow.restore()})
ipcMain.on('close', () => {mainWindow.close(); ipcMain.emit('login_exit');})
